File: source/com/barrybecker4/simulation/reactiondiffusion/algorithm/GrayScottModel.java
/** Copyright by Barry G. Becker, 2000-2011. Licensed under MIT License: http://www.opensource.org/licenses/MIT  */
package com.barrybecker4.simulation.reactiondiffusion.algorithm;

import java.awt.*;


/**
 * Data structure for the Gray-Scott algorithm.
 * @author Barry Becker.
 */
public final class GrayScottModel {

    public static final double K0 = 0.079;
    public static final double F0 = 0.02;

    private static final double INITIAL_U = 0.5;
    private static final double INITIAL_V = 0.25;

    /** concentrations of the 2 chemicals, u and v. */
    double[][] u;
    double[][] v;
    double[][] tmpU;
    double[][] tmpV;

    private double k_;
    private double f_;
    private double initialK_;
    private double initialF_;

    private int width_, height_;


    /**
     * Constructor
     * @param width width of computational space.
     * @param height height of computational space.
     */
    GrayScottModel(int width, int height) {
        this.width_ = width;
        this.height_ = height;
        this.initialF_ = F0;
        this.initialK_ = K0;
        resetState();
    }

    public double getU(int x, int y) {
        return u[x][y];
    }

    public double getV(int x, int y) {
        return v[x][y];
    }


    public int getWidth() {
        return width_;
    }

    public int getHeight() {
        return height_;
    }

    public void setSize(Dimension requestedNewSize) {
        width_ = requestedNewSize.width;
        height_ = requestedNewSize.height;
    }

    public void setF(double f) {
        f_ = f;
    }

    double getF() {
        return f_;
    }

    public void setK(double k) {
        k_ = k;
    }

    double getK() {
        return k_;
    }

    /**
     * Exchange the u, v fields with the tmp versions.
     */
    public void commitChanges() {
        double[][] temp = tmpU;
        tmpU = u;
        u = temp;

        temp = tmpV;
        tmpV = v;
        v = temp;
    }

    public double getNeighborSum(double tmp[][], int x, int y) {

       return tmp[x + 1][y]
                + tmp[x - 1][y]
                + tmp[x][y + 1]
                + tmp[x][y - 1];
    }

    public double getEdgeNeighborSum(double tmp[][], int x, int y) {

        return tmp[getPeriodicXValue(x + 1, width_)][y]
                  + tmp[getPeriodicXValue(x - 1, width_)][y]
                  + tmp[x][getPeriodicXValue(y + 1, height_)]
                  + tmp[x][getPeriodicXValue(y - 1, height_)];
    }

    /**
     * Create some initial pattern of chemical that represents the initial condition.
     */
    public void resetState()  {

        f_ = initialF_;
        k_ = initialK_;

        u = new double[width_][height_];
        v = new double[width_][height_];
        tmpU = new double[width_][height_];
        tmpV = new double[width_][height_];

        stampInitialSquare(0, 0, width_, height_, 1, 0);

        // random square 1
        int w3 = width_ / 3;
        int h3 = height_ / 3;
        stampInitialSquare(w3, h3, w3, h3, INITIAL_U, INITIAL_V);

        // random square 2
        int w7 = width_ / 7;
        int h5 = height_ / 5;
        stampInitialSquare(5*w7, 3*h5, w7, h5, INITIAL_U, INITIAL_V);
    }

    /**
     * Place a square of chemicals with the initial concentrations.
     */
    private void stampInitialSquare(int startX, int startY, int width, int height, double initialU, double initialV) {
        for (int x = 0; x < width; x++) {
            for (int y = 0; y < height; y++) {
                tmpU[startX + x][startY + y] = initialU;
                tmpV[startX + x][startY + y] = initialV;
            }
        }
    }

    /**
     * Periodic boundary conditions.
     * @return new x value taking into account wrapping boundaries.
     */
    private static int getPeriodicXValue(int x, int max) {
        int xp = x % max;
        return xp<0 ? xp + max : xp;
    }
}

File: scala-source/com/barrybecker4/simulation/reactiondiffusion/algorithm/GrayScottModel.scala
/** Copyright by Barry G. Becker, 2000-2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT  */
package com.barrybecker4.simulation.reactiondiffusion.algorithm

import GrayScottModel._
import java.awt._
import com.barrybecker4.simulation.reactiondiffusion.algorithm.configuration._


/**
  * Data structure for the Gray-Scott algorithm.
  * @author Barry Becker.
  */
object GrayScottModel {
  /** Feed rate */
  val F0: Double =  0.025

  /** Decay rate */
  val K0: Double =  0.079

  val H0: Double = 1.0

  /** Radius for area of effect when doing manual modification with click/drag brush */
  val DEFAULT_BRUSH_RADIUS = 2
  /** The amount of impact that the brush has. */
  val DEFAULT_BRUSH_STRENGTH = 0.3

  /** Periodic boundary conditions.
    * @return new x value taking into account wrapping boundaries.
    */
  def getPeriodicXValue(x: Int, max: Int): Int = {
    val xp = x % max
    if (xp < 0) xp + max else xp
  }
}

/**
  * @param width  width of computational space.
  * @param height height of computational space.
  */
final class GrayScottModel(var width: Int, var height: Int) extends InitializableGrid {

  var initializer: Initializer = Initializer.DEFAULT_INITIALIZER

  /** concentrations of the 2 chemicals, u and v. */
  var u: Array[Array[Double]] = _
  var v: Array[Array[Double]] = _

  private[algorithm] var tmpU: Array[Array[Double]] = _
  private[algorithm] var tmpV: Array[Array[Double]] = _

  private var k: Double = _
  private var f: Double = _
  resetState()

  def getWidth: Int = width
  def getHeight: Int = height

  def setSize(requestedNewSize: Dimension) {
    this.width = requestedNewSize.width
    this.height = requestedNewSize.height
  }

  def setInitializer(init: Initializer): Unit = {
    initializer = init
    resetState()
  }

  def setF(f: Double) { this.f = f }
  private[algorithm] def getF = f

  def setK(k: Double) {this.k = k}
  private[algorithm] def getK = k

  /** Exchange the u, v fields with the tmp versions.  */
  private[algorithm] def commitChanges()  {
    var temp = tmpU
    tmpU = u
    u = temp
    temp = tmpV
    tmpV = v
    v = temp
  }

  private[algorithm] def getNeighborSum(tmp: Array[Array[Double]], x: Int, y: Int): Double =
    tmp(x + 1)(y) +
      tmp(x - 1)(y) +
      tmp(x)(y + 1) +
      tmp(x)(y - 1)

  private[algorithm] def getEdgeNeighborSum(tmp: Array[Array[Double]], x: Int, y: Int): Double =
    tmp(getPeriodicXValue(x + 1, width))(y) +
      tmp(getPeriodicXValue(x - 1, width))(y) +
      tmp(x)(getPeriodicXValue(y + 1, height)) +
      tmp(x)(getPeriodicXValue(y - 1, height))

  /** Create some initial pattern of chemical that represents the initial condition. */
  private[algorithm] def resetState(): Unit = {
    f = F0
    k = K0

    u = Array.ofDim[Double](width, height)
    v = Array.ofDim[Double](width, height)
    tmpU = Array.ofDim[Double](width, height)
    tmpV = Array.ofDim[Double](width, height)

    initializer.initialize(this, width, height)
    commitChanges()
  }

  def initializePoint(x: Int, y: Int, initialU: Double, initialV: Double): Unit = {
    if (x >= 0 && x < width && y >= 0 && y < height) {
      tmpU(x)(y) = initialU
      tmpV(x)(y) = initialV
      u(x)(y) = initialU
      v(x)(y) = initialV
    }
  }
}
